### ISSUE TITLE
Wiping out the target namespace

### ISSUE THREAD

**Issue by aplteam** [the-carlisle-group/Acre-Desktop#62]
As Phil pointed out the other day what I do with all the APLTree projects is actually packet management.

I have a file publish.config (a misnomer, I know) which, among other things, defines dependencies.

This is mainly about scripts. For example, `CompareSimple` requires `APLTreeUtils`. It includes it with 

```
:Include APLTreeUtils
````

so I have to make sure that it is a sibling of `CompareSimple`.

All my project follow the same layout: for a project `Foo` I have a namespace `#._Foo` which contains at the very least

```
Foo
TestCases
```

Now in order to establish a project I use my own user command  `]OpenMyAcreProject`.

This command takes the name of the project (`Foo`) and creates a namespace `#._Foo`.

It then loads all the scripts needed according to publish.config into `#`.

It then establishes references (!) in `#._Foo` with, say:

```
#._Foo.APLTreeUtils←#.APLTreeUtils
```

That way I solved all dependency problem under acre 3. 

The approach has three disadvantages:

1. It requires a certain directory structure because I do no support downloading and unzipping of such dependencies (yet).

2. It does not support version numbers. A dependency just means "get me the latest version".

3. **It does _not_ work at all under acre 4.**

The reason is that acre 4 first wipes out the content of `#._Foo`, so my references all disappear, and the attempt of acre 4 to fix a script that carries  `:Include APLTreeUtils` therefore fails.

I discussed the issue with Phil, and we believe that the behaviour was changed for two reasons:

1. Disallow nested open projects. This must be achieved, but it can be done by other means than expunging everything: just "close" any nested project.
1. Get rid of the performance penalty that occurred when large applications like FlipDB or CAS were opened twice. This cannot be helped, but instead one could `}OFF` and start a fresh session. That does not need much effort, and it is still way faster.

Therefore I propose that we bring back acre's old behaviour of leaving any target namespace alone and just open/load a project into it, no matter what's already there.

**Comment by PaulMansour** [comment]
Would you be open to re-evaluating your package management system in light of Acre 4?

Acre 4 gives us the opportunity I think to have a very simple, consistent package management system. I have a rough implementation already. Be happy to demo this next week.

In the meantime, a specific question: what are "scripts"? Do you just mean an acre project? Or some other entity?

**Comment by aplteam** [comment]
> Would you be open to re-evaluating your package management system in light of Acre 4?

Of course!

But regardless of how I organize them, we should carefully consider what to do with the contents of a target namespace.

> In the meantime, a specific question: what are "scripts"? Do you just mean an acre project? Or some other entity?

Classes and scripted namespaces.

**Comment by e9gille** [comment]
This sounds like one of the suggestions I made to inject references to dependencies. In short, if acre ends up managing the loading of dependencies I guess it doesn't matter if it wipes the target namespace first.

**Comment by PaulMansour** [comment]
> Classes and scripted namespaces.

Do you mean projects that just happen to be single classes or scripted namespaces?

**Comment by PaulMansour** [comment]
> But regardless of how I organize them, we should carefully consider what to do with the contents of a target namespace.

I think it is a bad design to rely on stuff in the project namespace surviving ]Acre.Open.  Its an invitation for bugs.

**Comment by PaulMansour** [comment]
> This sounds like one of the suggestions I made to inject references to dependencies. In short, if acre ends up managing the loading of dependencies I guess it doesn't matter if it wipes the target namespace first.

I think no matter how dependencies are managed, by Acre or an outside app, it is a bad idea not to wipe out the target namespace on load or open.

**Comment by PaulMansour** [comment]
By definition, in Acre 4, a project cannot BE a scripted namespace or class, though it may CONTAIN these.  

Is this a correct statement?

**Comment by e9gille** [comment]
> I think no matter how dependencies are managed, by Acre or an outside app, it is a bad idea not to wipe out the target namespace on load or open.

Sure, what I meant is that it is a moot point. Kai raised it because he built his own dependency loader on top of acre. If dependencies are managed by acre then it is no longer relevant.

**Comment by e9gille** [comment]
> By definition, in Acre 4, a project cannot BE a scripted namespace or class, though it may CONTAIN these.

I meant to ask that actually. Maybe a project can't, could a package?

**Comment by aplteam** [comment]
Well, as I said, even if the project as such is a single class or namespace script, there is at least  a sibling namespace with the test cases. 

For APLTreeUtils for example:

```
#._APLTreeUtils
   APLTreeUtils
   TestCases
```
This is the absolute minimum.
If acre gives me the means to do it differently then fine.

**Comment by aplteam** [comment]
> I meant to ask that actually. Maybe a project can't, could a package?

Certainly! APLTreeUtils is just a scripted namespace, nothing else needed.

**Comment by aplteam** [comment]
Just to be clear: if you have the classes, A, AB and C with AB inheriting
from A and both A and AB calling methods in C you are proposing that after
loading we have something like this in the WS:

#.ns1.A
#.ns2 .AB
#.ns3.C

??

**Comment by e9gille** [comment]
What I am saying is that if A, AB and C are delivered as one package then they would all end up in one namespace. If they are all independent packages then yes, each one in its own namespace.

In your example, when working on project AB with A and C as a dependencies:
```
#.APLTeam.ProjectA.v_0_1_0.A
#.APLTeam.ProjectC.v_1_0_0.C
#.ProjectAB.A←#.APLTeam.ProjectA.v_0_1_0
#.ProjectAB.C←#.APLTeam.ProjectC.v_1_0_0
#.ProjectA.AB
```
With AB:
```
:Class AB: ##.A.A
    :Field Private Shared C←##.C.C
    foo←{
        C.goo ⍵
    }
...
:EndClass
```

I haven't worked out the details, but in general, to make package dependencies work I think we may have to go that route.

**Comment by PaulMansour** [comment]
I agree that a package should always be a tradition namespace.  If it is not, all kinds of special code is required when building and loading packages that contain one class or scripted namespaces.

I don't think I understand the problem of cross project referencing with respect to Kai's original question. Leaving versions aside for now if Kai has two projects as above, they will be loaded as so:
  
#.APLTeam.CompareSimple
#.APLTeam.APLTreeUtils

These are trad namespaces however, and if the first is a single class, and the second a single scripted namespace, then the ns tree continue on down with this:

#.APLTeam.CompareSimple.TheClass
#.APLTeam.APLTreeUtils.TheNS

In the case of scripted namespaces, I'm not there really is any need them, and it might be better to simply change the  project  to  bunch of functions, and get rid of the extra level.

Regardless, the question is then how :include TheNS in TheClass. Is it not simply, at worse case, with no extra references:

~~~
:Include ##.##.APLTreeUtils.TheNS
~~~

or just the fqp:

~~~
:Include #.APLTeam.APLTreeUtils.TheNS
~~~

Now the question is, is there a way to make this nicer, easier to read, shorter to write, and easier to change.

For a general approach, the best we can do is I think is having a reference to APLTreeUtils inside CompareSimple (or inside every trad namespace in the package/project:

~~~
:Include ##.APLTreeUtils.TheNS
~~~

We still need the ##, to step up out of the class, and the APLTreeUtils.TheNS because of the extra level  that a scripted namespaces imposes. Get rid of the script and we would have:

~~~
:Include ##.APLTreeUtils
~~~

:Include may be  a special problem  with respect to fixing the projects in the ws... not sure.

**Comment by e9gille** [comment]
@PaulMansour That is exactly what I was trying to say. Versioning is then potentially adding another nesting level.

**Comment by aplteam** [comment]
Honestly, does that look and feel right?!

If that's what package management is then I have a very strong urge to shy
away from it.

**Comment by e9gille** [comment]
I'm open to suggestions. Keeping in mind we need relocatable packages that will optionally need access to dependencies and resources, what is a better way forward?

**Comment by aplteam** [comment]
If I would know how to do it better I would come forward. I don't.

But that doesn't change my feeling that this is terrible.

**Comment by e9gille** [comment]
Kai, is your concern the deep nesting or the cluttering of the ws?

Having mulled this over during the day I've reached this conclusion:
* Distribution code is a dws with one object:
  * a trad ns
  or
  * a scripted class/namespace
* content is loaded into a nested path to distinguish versions of same package but allow code to be shared. How this is organised is a different matter. Could be anything between:
  * #.[package name] (only allow one version of the same package)
  * #.APM_Modules.[organisation].[package name].[version]
* Metadata is stored in the same space (project folder for access to resources etc.)
* A shortcut is created in the ancestor (consumer of the package) directly to the object.

So to wrap it up with an example, MyApp depends on HttpCommand (class) depends on Conga (trad ns):
HttpCommand required, so space prepared:
```
#.APM.HttpCommand.v2_0_0
#.APM.HttpCommand.v2_0_0.Meta - space with package info and project folder path
```
It depends on Conga, so is loaded first:
```
#.APM.Conga.v3_0_1
#.APM.Conga.v3_0_1.Meta
#.APM.Conga.v3_0_1.Conga - actual ns of code
```
Reference added to ancestor
`#.APM.HttpCommand.v2_0_0.Conga←#.APM.Conga.v3_0_1.Conga`
source loaded in:
`#.APM.HttpCommand.v2_0_0.HttpCommand`
If script, refix.
Finally, create ref in ancestor, which in top level is root:
`#.HttpCommand←#.APM.HttpCommand.v2_0_0.HttpCommand`

If acre loads a project with dependencies on packages, it can simply add the same refs to the project space once apm is done:
`#.MyApp.HttpCommand←#.HttpCommand`

Now if in MyApp I use HttpCommand I find it in the same space. When Conga needs access to the dlls on initialisation, it can simply refer to the parent:
`##.Meta.ProjectFolder`

I think this could work independently of acre. Also, the package configuration file could be shared (acre could use a superset of what is needed for apm). The only suggestion here would be to stick with one format for the config file that is easily modified via user commands (to add/remove dependencies from user commands).

How much have I missed?

**Comment by PaulMansour** [comment]
Kai, can you give some specific complaints, drawbacks, or inadequacies of what was originally described above?

**Comment by PaulMansour** [comment]
Also, think the issue here is somewhat if not totally tangential to packages, is it not? In any single, large project there will be many namespaces, classes, :Include statements, etc. There will be calls from one space to another, classes have to know about each other, etc. How is this to be done? How is it done now? Is the best technique for doing this any different than the best technique for dealing with multiple packages?

**Comment by aplteam** [comment]
When a class `Foo` needs `APLTreeUtils` then

```
:Class Foo
:Include APLTreeUtils
...
````
does the trick.

My "only" complaint is that it looks over-complicated, certainly way more complicated compared to what I've done so far. So it's just my guts that are revolting.

Have to read Gil's latest post now.

**Comment by e9gille** [comment]
> certainly way more complicated compared to what I've done so far.

What you did in CodeLib wasn't much different was it? But there we don't support multi versions of same package and the project folder was fixed.

**Comment by aplteam** [comment]
I must admit that I have just a faint memory after so many years but it is
not recursive, right?

**Comment by e9gille** [comment]
It is recursive and allows specifying version, but not sure what it does if there's a conflict. Packages are brought in flat in the root, so only one version can be allowed.
